src/io_callbacks.zig

```diff
@@ -0,0 +1,902 @@
+        assert(self.completed.out == null);
+    }
+
+    fn flush_completions(self: *IO, wait: bool) !void {
+        var cqes: [256]io_uring_cqe = undefined;
+        var wait_nr: u32 = if (wait) 1 else 0;
+        while (true) {
+            // Guard against waiting indefinitely (if there are too few requests inflight),
+            // especially if this is not the first time round the loop:
+            wait_nr = std.math.min(self.submitted, wait_nr);
+            const completed = self.ring.copy_cqes(&cqes, wait_nr) catch |err| switch (err) {
+                error.SignalInterrupt => continue,
+                else => return err,
+            };
+            self.submitted -= completed;
+            for (cqes[0..completed]) |cqe| {
+                const completion = @intToPtr(*Completion, @intCast(usize, cqe.user_data));
+                completion.result = cqe.res;
+                completion.next = null;
+                // We do not run the completion here (instead appending to a linked list):
+                // * to avoid recursion through `flush_submissions()` and `flush_completions()`,
+                // * to avoid unbounded stack usage, and
+                // * to avoid confusing stack traces.
+                self.completed.push(completion);
+            }
+            if (completed < cqes.len) break;
+        }
+    }
+
+    fn flush_submissions(self: *IO, wait: bool) !void {
+        var wait_nr: u32 = if (wait) 1 else 0;
+        while (true) {
+            wait_nr = std.math.min(self.queued + self.submitted, wait_nr);
+            _ = self.ring.submit_and_wait(wait_nr) catch |err| switch (err) {
+                error.SignalInterrupt => continue,
+                // Wait for some completions and then try again:
+                // See https://github.com/axboe/liburing/issues/281 re: error.SystemResources.
+                // Be careful also that copy_cqes() will flush before entering to wait (it does):
+                // https://github.com/axboe/liburing/commit/35c199c48dfd54ad46b96e386882e7ac341314c5
+                error.CompletionQueueOvercommitted, error.SystemResources => {
+                    try self.flush_completions(true);
+                    continue;
+                },
+                else => return err,
+            };
+            self.submitted += self.queued;
+            self.queued = 0;
+            break;
+        }
+    }
+
+    fn enqueue(self: *IO, completion: *Completion) void {
+        const sqe = self.ring.get_sqe() catch |err| switch (err) {
+            error.SubmissionQueueFull => {
+                completion.next = null;
+                self.unqueued.push(completion);
+                return;
+            },
+        };
+        self.queued += 1;
+        completion.prep(sqe);
+    }
+
+    /// This struct holds the data needed for a single io_uring operation
+    pub const Completion = struct {
+        io: *IO,
+        result: i32 = undefined,
+        next: ?*Completion = undefined,
```

Comment:
I think we should set the default value for `next` to `null` as soon as possible since we know that will be the initial value and we're not waiting for the value to be defined later. This would also make line 140 less of an issue if it was accidentally left out. We could then rather assert on line 140: `assert(completion.next == null)`.

Reply:
Setting the default value to `null` wouldn't be sufficient to avoid setting null explicitly on line 140 as completions may be re-used without re-initialization in some cases if an operation fails with EINTR.

What we could do here is unconditionally set `next` to null in `FIFO.push()`

Reply:
I think it's good that `FIFO.push()` asserts `next` is `null`, rather than unconditionally setting it, because that would help surface this kind of reuse without correctly setting `next` back to `null`, and we can then make sure that all our calls to `FIFO.push()` are already getting it right.

Let's set the default to `null` because that's a valid initial state, and then also make sure we reset `next` back to `null` as soon as the `next` pointer is no longer valid, which is the moment when it's shifted out a queue. `next` should either be `null` or a valid pointer if it's in a FIFO queue, but as soon as it's no longer in a queue it should be set back to `null`. So this is still an unconditional reset, but at the moment it's shifted rather than pushed.

Reply:
Implemented in the latest commit.